### PR TITLE
fix for issue #120, the probable root cause for PHP bug 68549

### PR DESCRIPTION
**Pull request by ComaVN**
fix for #120, and https://bugs.php.net/bug.php?id=68549

### CODE REVIEW

**Comment by ComaVN** [comment]
FYI, when I copy this to php's master, I get this:
```
=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Bug #51096 (Test for "first day" vs "first day of") [ext/date/tests/bug51096.phpt]
=====================================================================
roel@roelpad:~/git/php-src$ git status
```
but this seems unrelated, and I suspect it's a problem with my `re2c` while generating `parse_date.c` or something.

**Comment by ComaVN** [comment]
I rebased on the latest `master`, but had to fix a missing `#include`

**Comment by derickr** [comment]
I've merged this after rebasing, and refactoring, and the fix will be in PHP 8.1 and later once I make a timelib release and merge it into PHP.

File: tm2unixtime.c [hunk 1]
```diff
@@ -382,22 +382,51 @@ static void do_adjust_timezone(timelib_time *tz, timelib_tzinfo *tzi)
...
 			current = timelib_get_time_zone_info(tz->sse, tzi);
 			after = timelib_get_time_zone_info(tz->sse - current->offset, tzi);
+			actual_offset = after->offset;
+			actual_transition_time = after->transition_time;
+			if (current->offset == after->offset && tz->have_zone) {
+				/* Make sure we're not missing a DST change because we don't know the actual offset yet */
+				if (current->offset >= 0 && tz->dst && !current->is_dst) {
+						/* Timezone or its DST at or east of UTC, so the local time, interpreted as UTC, leaves DST (as defined in the actual timezone) before the actual local time */
+						timelib_time_offset *earlier;
+						earlier = timelib_get_time_zone_info(tz->sse - current->offset - 7200, tzi);
```
[the remark this answers is not in the document]
Comment (ComaVN) [reply tm2unixtime.c:401]: From [wikipedia](https://en.wikipedia.org/wiki/Daylight_saving_time):
> A one-hour change is usual, but twenty-minute and two-hour changes have been used in the past

I take this to mean that there's no point in looking further than 2*60*60 to detect upcoming switches.

A cleaner solution, without "magic", might be to extend the `timelib_time_offset` struct to also include `next_transition_time`, but I did not look into how to implement that

**Pull request closed without merging**